File: 2. Structural Design Patterns/Decorator/Go/main.go
package main

import "fmt"

// Decorator Pattern Example in Go
// Notifier: Basic + Email + SMS + Slack decorators

type Notifier interface{ Send(msg string) }

type BasicNotifier struct{}
func (BasicNotifier) Send(msg string){ fmt.Println("Base notify:", msg) }

type Decorator struct{ wrap Notifier }
func (d Decorator) Send(msg string){ d.wrap.Send(msg) }

type EmailDecorator struct{ Decorator }
func (e EmailDecorator) Send(msg string){ fmt.Println("Email sent:", msg); e.Decorator.Send(msg) }

type SMSDecorator struct{ Decorator }
func (s SMSDecorator) Send(msg string){ fmt.Println("SMS sent:", msg); s.Decorator.Send(msg) }

type SlackDecorator struct{ Decorator }
func (s SlackDecorator) Send(msg string){ fmt.Println("Slack message sent:", msg); s.Decorator.Send(msg) }

func main(){
	var n Notifier = BasicNotifier{}
	n = EmailDecorator{Decorator{n}}
	n = SMSDecorator{Decorator{n}}
	n = SlackDecorator{Decorator{n}}
	n.Send("System maintenance at 2 AM")
}

File: 2. Structural Design Patterns/Decorator/Rust/main.rs
// Decorator Pattern Example in Rust
// Notifier: Basic + Email + SMS + Slack decorators

trait Notifier { fn send(&self, message: &str); }

struct BasicNotifier;
impl Notifier for BasicNotifier { fn send(&self, message: &str){ println!("Base notify: {}", message); } }

struct EmailDecorator<T: Notifier> { inner: T }
impl<T: Notifier> Notifier for EmailDecorator<T> { fn send(&self, message: &str){ println!("Email sent: {}", message); self.inner.send(message); } }

struct SMSDecorator<T: Notifier> { inner: T }
impl<T: Notifier> Notifier for SMSDecorator<T> { fn send(&self, message: &str){ println!("SMS sent: {}", message); self.inner.send(message); } }

struct SlackDecorator<T: Notifier> { inner: T }
impl<T: Notifier> Notifier for SlackDecorator<T> { fn send(&self, message: &str){ println!("Slack message sent: {}", message); self.inner.send(message); } }

fn main(){
    let n = SlackDecorator { inner: SMSDecorator { inner: EmailDecorator { inner: BasicNotifier } } };
    n.send("System maintenance at 2 AM");
}
